azurerm/resource_arm_network_watcher_flow_log.go

```diff
@@ -0,0 +1,387 @@
+	}
+
+	if props := fli.FlowLogProperties; props != nil {
+		d.Set("enabled", props.Enabled)
+
+		// Azure API returns "" when flow log is disabled
+		// Don't overwrite to prevent storage account ID diff when that is the case
+		if *props.StorageID != "" {
+			d.Set("storage_account_id", props.StorageID)
+		}
+
+		if err := d.Set("retention_policy", flattenAzureRmNetworkWatcherFlowLogRetentionPolicy(props.RetentionPolicy)); err != nil {
+			return fmt.Errorf("Error setting `retention_policy`: %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func resourceArmNetworkWatcherFlowLogDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).Network.WatcherClient
+	ctx, cancel := timeouts.ForDelete(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	id, err := ParseNetworkWatcherFlowLogID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	// Get current flow log status
+	statusParameters := network.FlowLogStatusParameters{
+		TargetResourceID: &id.NetworkSecurityGroupID,
+	}
+	future, err := client.GetFlowLogStatus(ctx, id.ResourceGroup, id.NetworkWatcherName, statusParameters)
+	if err != nil {
+		return fmt.Errorf("Error getting Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", id.NetworkSecurityGroupID, id.NetworkWatcherName, id.ResourceGroup, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for retrieval of Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", id.NetworkSecurityGroupID, id.NetworkWatcherName, id.ResourceGroup, err)
+	}
+
+	fli, err := future.Result(*client)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", id.NetworkSecurityGroupID, id.NetworkWatcherName, id.ResourceGroup, err)
+	}
+
+	// There is no delete in Azure API. Disabling flow log is effectively a delete in Terraform.
+	if *fli.FlowLogProperties.Enabled {
+		fli.FlowLogProperties.Enabled = utils.Bool(false)
+		fli.FlowAnalyticsConfiguration = nil
+
+		setFuture, err := client.SetFlowLogConfiguration(ctx, id.ResourceGroup, id.NetworkWatcherName, fli)
+		if err != nil {
+			return fmt.Errorf("Error disabling Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", id.NetworkSecurityGroupID, id.NetworkWatcherName, id.ResourceGroup, err)
+		}
+
+		if err = setFuture.WaitForCompletionRef(ctx, client.Client); err != nil {
+			return fmt.Errorf("Error waiting for completion of disabling Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", id.NetworkSecurityGroupID, id.NetworkWatcherName, id.ResourceGroup, err)
+		}
+	}
+
+	return nil
+}
+
+func expandAzureRmNetworkWatcherFlowLogRetentionPolicy(d *schema.ResourceData) *network.RetentionPolicyParameters {
+	vs := d.Get("retention_policy").([]interface{})
+	v := vs[0].(map[string]interface{})
```

Comment:
More crashes have been occurring because we aren't nil checking lists/sets when casting them to `map[string]interface{}`. I don't think it'll crash here because every value is required and non-empty but it might be best practice to just nil check these anyway before casting to `map[string]interface{}`

Reply:
this is handled above 👍